File: src/ruby/tormenta20/models/arma.rb
# frozen_string_literal: true

require "active_record"
require "json"

module Tormenta20
  module Models
    # Model representing an Arma (Weapon) in Tormenta20.
    #
    # Weapons are categorized by type (simple, martial, exotic, firearm)
    # and can be either melee or ranged.
    #
    # @example Find all martial weapons
    #   Tormenta20.armas.marciais
    #
    # @example Find ranged weapons
    #   Tormenta20.armas.ranged
    #
    # @example Find weapons by damage type
    #   Tormenta20.armas.by_damage_type("corte")
    #
    # @!attribute [rw] id
    #   @return [String] Unique identifier (e.g., "espada_longa")
    # @!attribute [rw] name
    #   @return [String] Weapon name in Portuguese
    # @!attribute [rw] category
    #   @return [String] Weapon category: "simples", "marciais", "exoticas", or "fogo"
    # @!attribute [rw] price
    #   @return [String] Price in Tormenta currency
    # @!attribute [rw] damage
    #   @return [String] Damage dice (e.g., "1d8")
    # @!attribute [rw] damage_type
    #   @return [String] Type of damage: "corte", "perfuracao", "impacto"
    # @!attribute [rw] critical
    #   @return [String] Critical hit modifier (e.g., "19/x2")
    # @!attribute [rw] range
    #   @return [String, nil] Range for ranged weapons, nil for melee
    # @!attribute [rw] weight
    #   @return [String] Weight in slots/spaces
    # @!attribute [rw] properties
    #   @return [Array<String>] Special weapon properties
    # @!attribute [rw] description
    #   @return [String, nil] Optional description
    class Arma < Base
      self.table_name = "armas"

      # @return [Array<String>] Valid weapon categories
      CATEGORIES = %w[simples marciais exoticas fogo].freeze

      validates :id, presence: true, uniqueness: true
      validates :name, presence: true
      validates :category, presence: true, inclusion: { in: CATEGORIES }

      # @!method simples
      #   Filter simple weapons only.
      #   @return [ActiveRecord::Relation<Arma>]
      scope :simples, -> { where(category: "simples") }

      # @!method marciais
      #   Filter martial weapons only.
      #   @return [ActiveRecord::Relation<Arma>]
      scope :marciais, -> { where(category: "marciais") }

      # @!method exoticas
      #   Filter exotic weapons only.
      #   @return [ActiveRecord::Relation<Arma>]
      scope :exoticas, -> { where(category: "exoticas") }

      # @!method fogo
      #   Filter firearms only.
      #   @return [ActiveRecord::Relation<Arma>]
      scope :fogo, -> { where(category: "fogo") }

      # @!method by_category(category)
      #   Filter weapons by category.
      #   @param category [String] Category to filter by
      #   @return [ActiveRecord::Relation<Arma>]
      scope :by_category, ->(c) { where(category: c) }

      # @!method by_damage_type(type)
      #   Filter weapons by damage type.
      #   @param type [String] Damage type: "corte", "perfuracao", or "impacto"
      #   @return [ActiveRecord::Relation<Arma>]
      scope :by_damage_type, ->(t) { where(damage_type: t) }

      # @!method ranged
      #   Filter ranged weapons only (weapons with a range value).
      #   @return [ActiveRecord::Relation<Arma>]
      scope :ranged, -> { where.not(range: nil) }

      # @!method melee
      #   Filter melee weapons only (weapons without a range value).
      #   @return [ActiveRecord::Relation<Arma>]
      scope :melee, -> { where(range: nil) }

      # Check if this weapon is ranged.
      #
      # @return [Boolean] true if the weapon has a range value
      def ranged?
        !range.nil?
      end

      # Convert the weapon to a hash representation.
      #
      # @return [Hash] Hash with all weapon attributes (nil values excluded)
      def to_h
        {
          id: id,
          name: name,
          category: category,
          price: price,
          damage: damage,
          damage_type: damage_type,
          critical: critical,
          range: range,
          weight: weight,
          properties: properties,
          description: description
        }.compact
      end
    end
  end
end

File: src/ts/models/arma.ts
import { BaseModel } from "./base.js"
import { Query } from "../query.js"
import { getBookReference, type BookReference } from "./concerns/book_referenceable.js"

const TABLE = "armas"

/** Proficiency category of a weapon. */
export type ArmaCategory = "simples" | "marciais" | "exoticas" | "fogo"

/** Physical damage type dealt by a weapon. */
export type DamageType = "corte" | "perfuracao" | "impacto"

/**
 * An Arma (weapon) in Tormenta20.
 *
 * @example
 * ```ts
 * Arma.marciais().all()
 * Arma.ranged().all()
 * Arma.melee().where("damage_type = ?", "corte").all()
 * Arma.find("espada_longa")
 * ```
 */
export class Arma extends BaseModel {
  get category(): ArmaCategory { return this._row.category as ArmaCategory }
  /** Price in tibares. */
  get price(): number { return (this._row.price as number) ?? 0 }
  /** Damage dice expression (e.g. `"1d8"`). */
  get damage(): string { return this._row.damage as string }
  get damageType(): DamageType { return this._row.damage_type as DamageType }
  /** Critical threat range and multiplier (e.g. `"19-20/x2"`). */
  get critical(): string { return this._row.critical as string }
  /** Range increment in metres, or `null` for melee-only weapons. */
  get range(): string | null { return (this._row.range as string) || null }
  /** Weight in spaces (espaços). */
  get weight(): number { return (this._row.weight as number) ?? 0 }
  /** JSON array of special property objects. */
  get properties(): unknown[] { return BaseModel.json(this._row.properties, []) }
  get description(): string | null { return (this._row.description as string) ?? null }

  /** `true` if the weapon has a range increment. */
  get isRanged(): boolean { return this.range !== null }
  /** `true` if the weapon is melee-only. */
  get isMelee(): boolean { return !this.isRanged }

  /** Returns the sourcebook page reference, or `null` if not indexed. */
  bookReference(): BookReference | null { return getBookReference(this.id) }

  toH() {
    return {
      id: this.id, name: this.name, category: this.category, price: this.price,
      damage: this.damage, damage_type: this.damageType, critical: this.critical,
      range: this.range, weight: this.weight, properties: this.properties,
      description: this.description,
    }
  }

  static query(): Query<Arma> { return BaseModel.makeQuery(TABLE, Arma) }
  static all(): Arma[] { return Arma.query().all() }
  static find(id: string): Arma | null { return Arma.query().find(id) }
  static first(): Arma | null { return Arma.query().first() }
  static last(): Arma | null { return Arma.query().last() }
  static count(): number { return Arma.query().count() }

  static byCategory(cat: ArmaCategory): Query<Arma> { return Arma.query().where("category = ?", cat) }
  static byDamageType(dt: DamageType): Query<Arma> { return Arma.query().where("damage_type = ?", dt) }
  /** Returns weapons with a range increment. */
  static ranged(): Query<Arma> { return Arma.query().where("range IS NOT NULL AND range != ''") }
  /** Returns melee-only weapons. */
  static melee(): Query<Arma> { return Arma.query().where("range IS NULL OR range = ''") }

  static simples(): Query<Arma> { return Arma.byCategory("simples") }
  static marciais(): Query<Arma> { return Arma.byCategory("marciais") }
  static exoticas(): Query<Arma> { return Arma.byCategory("exoticas") }
  static fogo(): Query<Arma> { return Arma.byCategory("fogo") }
}
